File: android/foxui/src/main/java/io/gitee/stevenzack/foxui/FObject/Widget/FWebView.java
package io.gitee.stevenzack.foxui.FObject.Widget;

import android.os.Build;
import android.support.v4.view.ViewCompat;
import android.util.Log;
import android.view.ViewGroup;
import android.webkit.DownloadListener;
import android.webkit.WebResourceRequest;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import org.json.JSONObject;
import org.json.JSONTokener;

import java.util.HashMap;
import java.util.Iterator;

import fox.Fox;
import io.gitee.stevenzack.foxui.FObject.FObject;
import io.gitee.stevenzack.foxui.FoxActivity;

public class FWebView extends FObject {
    public WebView v;

    public FWebView(FoxActivity foxActivity) {
        parentController = foxActivity;
        v = new WebView(parentController);
        v.setId(ViewCompat.generateViewId());
        view=v;
        v.getSettings().setDefaultTextEncodingName("UTF -8");
        parentController.onDestroyEvent.add(new Runnable() {
            @Override
            public void run() {
                if (v!=null) {
                    ViewGroup parent = (ViewGroup) v.getParent();
                    if (parent != null) {
                        parent.removeView(v);
                    }
                    v.stopLoading();
                    v.getSettings().setJavaScriptEnabled(false);
                    v.clearHistory();
                    v.clearView();
                    v.removeAllViews();
                    v.destroy();
                    v = null;
                }
                parentController.viewmap.remove(vid);
            }
        });
    }

    @Override
    public String getAttr(String attr) {
        return super.getAttr(attr);
    }

    @Override
    public String setAttr(String attr, final String value, final String value2) {
        String str = super.setAttr(attr, value, value2);
        if (str != null) {
            return str;
        }
        switch (attr) {
            case "Uri":
                v.loadUrl(value);
                break;
            case "Focusable":
                v.setFocusable(value.equals("true"));
                break;
            case "SupportZoom":
                v.getSettings().setSupportZoom(value.equals("true"));
                break;
            case "BuiltInZoomControls":
                v.getSettings().setBuiltInZoomControls(value.equals("true"));
                break;
            case "UseWideViewPort":
                v.getSettings().setUseWideViewPort(value.equals("true"));
                break;
            case "AllowContentAccess":
                v.getSettings().setAllowContentAccess(value.equals("true"));
                break;
            case "AllowFileAccess":
                v.getSettings().setAllowFileAccess(value.equals("true"));
                break;
            case "AllowFileAccessFromFileURLs":
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
                    v.getSettings().setAllowFileAccessFromFileURLs(value.equals("true"));
                }
                break;
            case "AllowUniversalAccessFromFileURLs":
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
                    v.getSettings().setAllowUniversalAccessFromFileURLs(value.equals("true"));
                }
                break;
            case "AppCacheEnabled":
                v.getSettings().setAppCacheEnabled(value.equals("true"));
                break;
            case "BlockNetworkImage":
                v.getSettings().setBlockNetworkImage(value.equals("true"));
                break;
            case "BlockNetworkLoads":
                v.getSettings().setBlockNetworkLoads(value.equals("true"));
                break;
            case "JavaScriptEnabled":
                v.getSettings().setJavaScriptEnabled(value.equals("true"));
                break;
            case "OffscreenPreRaster":
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                    v.getSettings().setOffscreenPreRaster(value.equals("true"));
                }
                break;
            case "SaveFormData":
                v.getSettings().setSaveFormData(value.equals("true"));
                break;
            case "SupportMultipleWindows":
                v.getSettings().setSupportMultipleWindows(value.equals("true"));
                break;
            case "TextZoom":
                try {
                    v.getSettings().setTextZoom(Integer.parseInt(value));
                } catch (Exception e) {
                    e.printStackTrace();
                }
                break;
            case "UserAgentString":
                v.getSettings().setUserAgentString(value);
                break;
            // -------------------
            case "HandleUrl":
                MyWebViewClient wc=new MyWebViewClient();
                if (!value.equals("")) {
                    try {
                        wc.urlHandlersMap = new HashMap<>();
                        JSONObject object = (JSONObject) (new JSONTokener(value).nextValue());
                        Log.d(TAG, "setAttr: HandleUrl"+value);
                        for (Iterator<String> it = object.keys(); it.hasNext(); ) {
                            String key = it.next();
                            wc.urlHandlersMap.put(key, object.getString(key));
                        }
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                }
                v.setWebViewClient(wc);
                break;
            case "DownloadListener":
                v.setDownloadListener(new DownloadListener() {
                    @Override
                    public void onDownloadStart(String url, String userAgent, String contentDisposition, String mimetype, long contentLength) {
                        Fox.triggerFunction(parentController,value, url,"","");
                    }
                });
                break;
            case "LoadHtmlData":
                String a = value.replace("#", "%23");
                v.loadData(a, "text/html; charset=UTF-8", null);
                break;
        }
        return "";
    }
    class MyWebViewClient extends WebViewClient {
        public HashMap<String,String> urlHandlersMap=null;
        @Override
        public boolean shouldOverrideUrlLoading(WebView view, WebResourceRequest request) {
            if (urlHandlersMap==null) {
                return super.shouldOverrideUrlLoading(view, request);
            }else {
                Log.d(TAG, "shouldOverrideUrlLoading: "+urlHandlersMap.size());
                if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.LOLLIPOP) {
                    String url = request.getUrl().toString();
                    if (urlHandlersMap.containsKey(removeQuery(url))) {
                        String fnId = urlHandlersMap.get(removeQuery(url));
                        String s= Fox.triggerFunction(parentController,fnId, url,"","");
                        if(s!=null&&s.equals("true"))
                            return true;
                    }else if (urlHandlersMap.containsKey(url)) {
                        String fnId = urlHandlersMap.get(url);
                        String s= Fox.triggerFunction(parentController,fnId, url,"","");
                        if(s!=null&&s.equals("true"))
                            return true;
                    }
                }
            }
            return false;
        }

        @Override
        public boolean shouldOverrideUrlLoading(WebView view, String url) {
            if (urlHandlersMap==null) {
                return super.shouldOverrideUrlLoading(view, url);
            }else {
                Log.d(TAG, "shouldOverrideUrlLoading2: "+urlHandlersMap.size());
                if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.LOLLIPOP) {
                    if (urlHandlersMap.containsKey(removeQuery(url))) {
                        String fnId = urlHandlersMap.get(removeQuery(url));
                        String s= Fox.triggerFunction(parentController,fnId, url,"","");
                        if(s!=null&&s.equals("true"))
                            return true;
                    }else if (urlHandlersMap.containsKey(url)) {
                        String fnId = urlHandlersMap.get(url);
                        String s= Fox.triggerFunction(parentController,fnId, url,"","");
                        if(s!=null&&s.equals("true"))
                            return true;
                    }
                }
            }
            return false;
        }
    }

    public static String removeQuery(String url) {
        int i = url.indexOf("?");
        if (i < 0) {
            return url;
        }
        return url.substring(0, i);
    }
}

File: fcore/widget/fwebview/fwebview.go
package fwebview

import (
	"github.com/StevenZack/fox/fcore"
	"github.com/StevenZack/fox/fcore/values/fcolor"
)

type FWebView struct {
	fcore.FBaseView
}

func New(a fcore.IActivity) *FWebView {
	f:=&FWebView{}
	f.A=a
	f.Vid=fcore.NewToken()
	f.VType="WebView"
	fcore.ViewMap.Set(f.Vid,f)
	f.A.NewObject(f.VType,f.Vid)
	fnId := f.Vid+":onclick"
	fcore.EventMap.Set(fnId, func(fcore.IActivity, string, string, string) string {
		if f.FnOnClick != nil {
			f.FnOnClick()
		}
		return ""
	})
	f.A.SetAttr(f.Vid, "OnClick", fnId, "")
	return f
}

func NewItem(a fcore.IActivity, uri string) *fcore.FBaseView {
	w := New(a)
	if len(uri) > 0 && uri[:1] == "/" {
		uri = "file://" + uri
	}
	return w.Focusable(false).LoadUri(uri).BackgroundColor(fcolor.Transparent).Size(-1, -1)
}
// --------------------------------------------------------

func (v *FWebView) LoadUri(s string) *FWebView {
	v.A.SetAttr(v.Vid, "Uri", s,"")
	return v
}
func (v *FWebView) Focusable(b bool) *FWebView {
	if b {
		v.A.SetAttr(v.Vid, "Focusable", "true","")
	} else {
		v.A.SetAttr(v.Vid, "Focusable", "false","")
	}
	return v
}
func (v *FWebView) SupportZoom(b bool) *FWebView {
	if b {
		v.A.SetAttr(v.Vid, "SupportZoom", "true","")
	} else {
		v.A.SetAttr(v.Vid, "SupportZoom", "false","")
	}
	return v
}
func (v *FWebView) BuiltInZoomControls(b bool) *FWebView {
	if b {
		v.A.SetAttr(v.Vid, "BuiltInZoomControls", "true","")
	} else {
		v.A.SetAttr(v.Vid, "BuiltInZoomControls", "false","")
	}
	return v
}
func (v *FWebView) UseWideViewPort(b bool) *FWebView {
	if b {
		v.A.SetAttr(v.Vid, "UseWideViewPort", "true","")
	} else {
		v.A.SetAttr(v.Vid, "UseWideViewPort", "false","")
	}
	return v
}
func (v *FWebView) AllowContentAccess(b bool) *FWebView {
	if b {
		v.A.SetAttr(v.Vid, "AllowContentAccess", "true","")
	} else {
		v.A.SetAttr(v.Vid, "AllowContentAccess", "false","")
	}
	return v
}
func (v *FWebView) AllowFileAccess(b bool) *FWebView {
	if b {
		v.A.SetAttr(v.Vid, "AllowFileAccess", "true","")
	} else {
		v.A.SetAttr(v.Vid, "AllowFileAccess", "false","")
	}
	return v
}
func (v *FWebView) AllowFileAccessFromFileURLs(b bool) *FWebView {
	if b {
		v.A.SetAttr(v.Vid, "AllowFileAccessFromFileURLs", "true","")
	} else {
		v.A.SetAttr(v.Vid, "AllowFileAccessFromFileURLs", "false","")
	}
	return v
}
func (v *FWebView) AllowUniversalAccessFromFileURLs(b bool) *FWebView {
	if b {
		v.A.SetAttr(v.Vid, "AllowUniversalAccessFromFileURLs", "true","")
	} else {
		v.A.SetAttr(v.Vid, "AllowUniversalAccessFromFileURLs", "false","")
	}
	return v
}
func (v *FWebView) AppCacheEnabled(b bool) *FWebView {
	if b {
		v.A.SetAttr(v.Vid, "AppCacheEnabled", "true","")
	} else {
		v.A.SetAttr(v.Vid, "AppCacheEnabled", "false","")
	}
	return v
}
func (v *FWebView) BlockNetworkImage(b bool) *FWebView {
	if b {
		v.A.SetAttr(v.Vid, "BlockNetworkImage", "true","")
	} else {
		v.A.SetAttr(v.Vid, "BlockNetworkImage", "false","")
	}
	return v
}
func (v *FWebView) BlockNetworkLoads(b bool) *FWebView {
	if b {
		v.A.SetAttr(v.Vid, "BlockNetworkLoads", "true","")
	} else {
		v.A.SetAttr(v.Vid, "BlockNetworkLoads", "false","")
	}
	return v
}
func (v *FWebView) JavaScriptEnabled(b bool) *FWebView {
	if b {
		v.A.SetAttr(v.Vid, "JavaScriptEnabled", "true","")
	} else {
		v.A.SetAttr(v.Vid, "JavaScriptEnabled", "false","")
	}
	return v
}
func (v *FWebView) OffscreenPreRaster(b bool) *FWebView {
	if b {
		v.A.SetAttr(v.Vid, "OffscreenPreRaster", "true","")
	} else {
		v.A.SetAttr(v.Vid, "OffscreenPreRaster", "false","")
	}
	return v
}
func (v *FWebView) SaveFormData(b bool) *FWebView {
	if b {
		v.A.SetAttr(v.Vid, "SaveFormData", "true","")
	} else {
		v.A.SetAttr(v.Vid, "SaveFormData", "false","")
	}
	return v
}
func (v *FWebView) SupportMultipleWindows(b bool) *FWebView {
	if b {
		v.A.SetAttr(v.Vid, "SupportMultipleWindows", "true","")
	} else {
		v.A.SetAttr(v.Vid, "SupportMultipleWindows", "false","")
	}
	return v
}
func (v *FWebView) TextZoom(i int) *FWebView {
	v.A.SetAttr(v.Vid, "TextZoom",fcore. SPrintf(i),"")
	return v
}
func (v *FWebView) UserAgentString(s string) *FWebView {
	v.A.SetAttr(v.Vid, "UserAgentString", s,"")
	return v
}
func (v *FWebView) HandleUrl(m map[string]func(string) bool) *FWebView {
	if m == nil {
		return v
	}
	ms := make(map[string]string)
	for k, h := range m {
		fnId :=fcore. NewToken()
		fcore.EventMap.Set(fnId, func(activity fcore.IActivity, s string, s2 string, s3 string) string {
			return fcore.SPrintf(h(s))
		})
		ms[k] = fnId
	}
	v.A.SetAttr(v.Vid, "HandleUrl", fcore.JsonObject(ms),"")
	return v
}
func (v *FWebView) OnDownload(f func(string)) *FWebView {
	fnId := fcore.NewToken()
	fcore.EventMap.Set(fnId, func(activity fcore.IActivity, s string, s2 string, s3 string) string {
		f(s)
		return ""
	})
	v.A.SetAttr(v.Vid, "DownloadListener", fnId,"")
	return v
}
func (v *FWebView) LoadHtmlData(s string) *FWebView {
	v.A.SetAttr(v.Vid, "LoadHtmlData", s,"")
	return v
}
